doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,510 @@
+```
+
+###### ROUTE_TABLE
+The existing ROUTE_TABLE for IPv4/IPv6 prefix routes is enhanced to accept an optional MPLS label stack component in the existing "nexthop" formatted-string attribute.  The new format of the nexthop attribute string with MPLS label stack is: "label0/.../labelN+ip-prefix".  For IP-only next hops, the previous string format is retained and no preceding label stack information will be included.
+
+``` rfc5234
+"ROUTE_TABLE":{{prefix}}
+    "nexthop":{{nexthop_list}}
+    "ifname":{{ifname_list}}
+
+; Defines schema for IPv4/IPv6 route table attributes
+key                         = ROUTE_TABLE:prefix       ; IPv4/IPv6 prefix
+; field                     = value
+nexthop                     = nexthop_list             ; List of nexthops.
+ifname                      = ifname_list              ; List of interfaces.
+;value annotations
+ifname                      = 1*64VCHAR                ; name of the Interface (Port Channel)
+```
+
+###### LABEL_ROUTE_TABLE
+A new LABEL_ROUTE_TABLE is introduced for MPLS in-segment entries to accept the same attributes as ROUTE_TABLE:
+- A "nexthop" formatted-string attribute with optional MPLS label stack component.
+- A "ifname" attribute.
+
+``` rfc5234
+"LABEL_ROUTE_TABLE":{{mpls_label}}
+    "nexthop":{{nexthop_list}}
+    "ifname":{{ifname_list}}
+
+; Defines schema for MPLS label route table attributes
+key                         = LABEL_ROUTE_TABLE:mpls_label ; MPLS label
+; field                     = value
+nexthop                     = STRING                   ; Comma-separated list of nexthops.
+ifname                      = STRING                   ; Comma-separated list of interfaces.
+```
+
+#### Software Modules
+This section describes modifications to SONiC infrastructure software modules to support MPLS.
+
+##### NetLink
+Netlink is an existing open source library imported to SONiC.
+Modifications to the existing NetLink MPLS implementation were needed to support MPLS attributes.
+###### Functions
+New accessors were added to retrieve the MPLS NH destination and TTL values:
+```
+  /* Accessor to retrieve MPLS destination */
+  extern struct nl_addr *	rtnl_route_nh_get_encap_mpls_dst(struct rtnl_nexthop *);
+  /* Accessor to retrieve MPLS TTL */
+  extern uint8_t		rtnl_route_nh_get_encap_mpls_ttl(struct rtnl_nexthop *);
+```
+
+##### IntfMgr
+IntfMgr is an existing daemon in SWSS container that monitors operations in CONFIG_DB on INTERFACE and PORTCHANNEL_INTERFACE tables.
+
+For MPLS, IntfMgr is modified to additionally process the "mpls" attribute and propagate this attribute to APPL_DB.
+
+##### FPM Syncd
+FPM Syncd is an existing daemon in BGP container that monitors NetLink messages from the SONiC routing stack for Route and Next Hop information.
+
+New support has been added to FPM Syncd for MPLS to process MPLS related Route and Next Hop information in the received NetLink messages and propagate this information to the APPL_DB.
```

Comment:
Juniper cRPD updates FPMsyncd with MPLS route information?

Reply:
Yes, that's right.